R/view-diff.R

```diff
@@ -41,12 +41,17 @@ diffviewer_widget <- function(old, new, width = NULL, height = NULL,
     bin_data <- read_raw(filename)
 
     # Assume .json and .download files are text
-    if (grepl("\\.json$", filename) || grepl("\\.download$", filename)) {
+    if (grepl("\\.(json|download|txt)$", filename)) {
       raw_to_utf8(bin_data)
     } else if (grepl("\\.png$", filename)) {
       paste0("data:image/png;base64,", jsonlite::base64_enc(bin_data))
     } else {
-      ""
+      # provide hash of file contents as a proxy to display binary differences
+      paste0(
+        "{shinytest} - [Spooky]Hash of file contents: ", digest::digest(bin_data, algo = "spookyhash"), "\n",
```

Comment:
How about using a more "standard" hash algorithm, like SHA1?

Reply:
Using the same setup as above, `sha1` is as slow as `md5`. Where as `spookyhash` is faster than `sha1`/`md5` for larger files.

```r
# "small" file
bench::mark(
  md5 = digest::digest(a),
  spooky = digest::digest(a, algo = "spookyhash"),
  sha1 = digest::digest(a, algo = "sha1"),
  check = FALSE
)
#> # A tibble: 3 × 13
#>   expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result memory     time      gc
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list> <list>     <list>    <list>
#> 1 md5          18.8µs   25.5µs    38118.    8.63KB    11.4   9997     3      262ms <NULL> <Rprofmem… <bench_t… <tibble …
#> 2 spooky         24µs   30.7µs    32743.    8.63KB     9.83  9997     3      305ms <NULL> <Rprofmem… <bench_t… <tibble …
#> 3 sha1         21.7µs   28.9µs    35311.    8.63KB    14.1   9996     4      283ms <NULL> <Rprofmem… <bench_t… <tibble …
```
```r
# 1000x sizeof("small" file)
bench::mark(
  md5 = digest::digest(b),
  spooky = digest::digest(b, algo = "spookyhash"),
  sha1 = digest::digest(b, algo = "sha1"),
  check = FALSE
)
#> # A tibble: 3 × 13
#>   expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result memory     time      gc
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list> <list>     <list>    <list>
#> 1 md5         212.8µs  262.2µs     3400.  101.19KB     6.43  1585     3      466ms <NULL> <Rprofmem… <bench_t… <tibble …
#> 2 spooky       30.1µs   36.1µs    25985.    8.63KB     7.80  9997     3      385ms <NULL> <Rprofmem… <bench_t… <tibble …
#> 3 sha1        216.7µs  236.6µs     3981.   90.73KB     8.63  1846     4      464ms <NULL> <Rprofmem… <bench_t… <tibble …
```

-------------------

I'm ok with using sha1 as it is familiar to users.

Reply:
Ah. It is micro seconds... Going to switch to `sha1`.